### ISSUE TITLE
Unclear definition of intervention

### ISSUE THREAD

**Issue by githubnemo** [reeceshuttle/intruder-dimensions#3]
Hey!

I'm currently trying to reproduce some of the effects (and mitigations) described in the paper to possibly integrate it in PEFT (related issue: https://github.com/huggingface/peft/issues/2907).

In the paper it is stated that we can do a weighted subtraction of the top-k intruder vectors:

<img width="980" height="222" alt="Image" src="https://github.com/user-attachments/assets/d40715b5-81a1-4ab5-8ee9-a40eed934a20" />

To me it is not entirely clear how do get the intruder dimensions `u_i s_i v_i`. I think there are two options

a) we use `SVD(W_0 + dW)`
b) we use `SVD(dW)`

I did some rudimentary testing on a single layer of a Llama-3.2 3B model trained on MetaMathQA (lora rank64) and found that using `W_0 + dW` seems to introduce more intruder dimensions than it removes them but I might have done something wrong (left unmodified `W_merged`, right "corrected" `W_merged` based on `SVD(W_0 + dW)`):

<img width="543" height="272" alt="Image" src="https://github.com/user-attachments/assets/14b3780b-e5f5-4734-8564-dc265f4ffde8" />

I'd appreciate if you could shed some light on this!

Link to notebook: https://gist.github.com/githubnemo/dbc8bbcea9f3266459fe54fd8216da69

**Comment by reeceshuttle** [comment]
@githubnemo interesting. ill take a more in depth look into this later, but to calculate intruder dimensions, we use option a) (`SVD(W_0+dW)`). 
One thing im noticing in your intruder dimension measurement is you are not taking the absolute value of the cosine similarities (the SVD, can sometimes return singular vectors in both U and V that are the negative of each other and therefore cancel out but would lead to cosine sim of -1 instead of 1 when compared to the pre-trained weights), which may be incorrectly flagging certain dimensions as intruder dimensions. See my implementation here: https://github.com/reeceshuttle/intruder-dimensions/blob/9b83bf208a124eb9d33f7a680d7924f1f1700953/find_intruder_dimensions.py#L12-L17

**Comment by reeceshuttle** [comment]
@githubnemo I tried to run your notebook but cannot due to version mismatch. could you send me the versions of the packages you are using for this notebook? or let me know the effect of the change i mention above (using `.abs()` before `.max`).

**Comment by githubnemo** [comment]
Hey @reeceshuttle, thanks for the reply!


> could you send me the versions of the packages you are using for this notebook?

Not sure what the issue is here, possibly the PEFT version because the checkpoint is from my machine and I'm using PEFT installed from source. So you might need to use the newest release or install PEFT from source.

>  or let me know the effect of the change i mention above (using `.abs()` before `.max`).

I've re-run the experiments using the following patch:

```diff
-    cos_sim = (U_lora64.T @ U_base).max(dim=1).values
+    cos_sim = (U_lora64.T @ U_base).abs().max(dim=1).values
```

The loss and accuracy results are, indeed, very different and are much closer to the paper. 

Visually it is still not what I would have expected from the paper but I'm testing with quite high threshold values, and using `dW` is less susceptible.

<img width="612" height="428" alt="Image" src="https://github.com/user-attachments/assets/0f0b008f-c182-4579-8c3a-43673ee984d5" />

<img width="612" height="428" alt="Image" src="https://github.com/user-attachments/assets/2e79740d-107b-4001-8cba-63b4a9299743" />

But testing forgetting via cross-entropy loss and task performance on accuracy yields the following results:

<img width="620" height="507" alt="Image" src="https://github.com/user-attachments/assets/e2cdecea-af01-434f-8cec-1d8ec9b020f2" />

<img width="620" height="507" alt="Image" src="https://github.com/user-attachments/assets/081b27ed-a3cf-4eca-88d6-365ae347322f" />

Regarding terminology:

- merged means computing cosine similarities and applying the mitigation on `W + dW` 
- merged-sign is the same but with using `abs()` for computing the similarities as per your suggestion
- dW means computing cosine similarities and applying the migitation on `dW`
- dW-sign is analogous to `merged-sign`

It seems to me that `merged-sign` is the best trade-off since it offers the best knowledge recovery while keeping close to the original performance with `lambda=0.5`. `dW-sign` seems to be ~equal to `merged-sign` but a bit off-set. At least this limited parameter search with a very, very limited forgetting test suggests that `merged-sign` is the way to go, so `W + dW` as a base for distances and mitigation and using `.abs()` in the cosine similarity computation.

Raw results:

```
RESULTS = {
    'dW': {
        'lambdas': [1.00, 0.75, 0.50, 0.25],
        'losses': [10.5743, 10.4626, 10.3382, 10.1989],
        'accuracies': [0.4874905231235785, 0.48142532221379836, 0.47687642153146326, 0.38741470811220624],
    },
    'dW-sign': {
        'lambdas': [1.00, 0.75, 0.50, 0.25],
        'losses': [10.5743, 10.5029, 10.4243, 10.3366],
        'accuracies': [0.4874905231235785, 0.48597422289613346, 0.5018953752843063, 0.4890068233510235],
    },
    'merged': {
        'lambdas': [1.00, 0.75, 0.50, 0.25],
        'losses': [10.5743, 10.6780, 10.8948, 10.8614],
        'accuracies': [0.4874905231235785, 0.48673237300985595, 0.45261561789234267, 0.37225170583775585],
    },
    'merged-sign': {
        'lambdas': [1.00, 0.75, 0.50, 0.25],
        'losses': [10.5743, 10.4615, 10.3259, 10.1092],
        'accuracies': [0.4874905231235785, 0.49052312357846856, 0.4829416224412434, 0.422289613343442],
    },
}
```

**Comment by reeceshuttle** [comment]
@githubnemo Thanks for your work on this! Exactly, using 'merged-sign' is the correct implementation (this is how this is implemented in our paper). I'm glad our results appear to replicate with your experiments.

> Visually it is still not what I would have expected from the paper

Could you clarify what the x and y axis are, what the left and right hand plots are, and add a legend of what color intensity relates to? With this info i can try to explain why the plots are expected or if they indeed are different.

**Comment by githubnemo** [comment]
> [@githubnemo](https://github.com/githubnemo) Thanks for your work on this! Exactly, using 'merged-sign' is the correct implementation (this is how this is implemented in our paper). I'm glad our results appear to replicate with your experiments.

Nice. Good to know. Then this issue is closed from my side. Thanks for 

> > Visually it is still not what I would have expected from the paper
> 
> Could you clarify what the x and y axis are, what the left and right hand plots are, and add a legend of what color intensity relates to? With this info i can try to explain why the plots are expected or if they indeed are different.

It would be simpler to look at https://gist.github.com/githubnemo/dbc8bbcea9f3266459fe54fd8216da69, cell 23 and 24 are basically the ones that generate the graphs.

**Comment by reeceshuttle** [comment]
Great. I'd say the left plot definitely makes sense- that 'empty' top row in the left plot will indicate an intruder dimension (low cosine similarity to the pre-trained singular vectors). I will say using a threshold of 0.9 is pretty high, so that may lead to more vectors being flagged than are actually 'intruder dimensions'. we generally used 0.5 in the paper but this is a hyperparameter of course.

Also, i just saw https://github.com/huggingface/peft/issues/2907. please let me know if i can support this implementation that you are doing in any way!
